File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/party/PartyDetailFragment.java
package com.habitrpg.android.habitica.ui.fragments.social.party;

import android.app.AlertDialog;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.facebook.drawee.view.SimpleDraweeView;
import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.components.AppComponent;
import com.habitrpg.android.habitica.data.InventoryRepository;
import com.habitrpg.android.habitica.data.SocialRepository;
import com.habitrpg.android.habitica.data.UserRepository;
import com.habitrpg.android.habitica.helpers.RxErrorHandler;
import com.habitrpg.android.habitica.models.inventory.Quest;
import com.habitrpg.android.habitica.models.inventory.QuestContent;
import com.habitrpg.android.habitica.models.social.Group;
import com.habitrpg.android.habitica.models.user.User;
import com.habitrpg.android.habitica.modules.AppModule;
import com.habitrpg.android.habitica.ui.activities.MainActivity;
import com.habitrpg.android.habitica.ui.fragments.BaseFragment;
import com.habitrpg.android.habitica.ui.fragments.inventory.items.ItemRecyclerFragment;
import com.habitrpg.android.habitica.ui.fragments.social.QuestDetailFragment;
import com.habitrpg.android.habitica.ui.helpers.DataBindingUtils;
import com.habitrpg.android.habitica.ui.helpers.MarkdownParser;
import com.habitrpg.android.habitica.ui.views.social.OldQuestProgressView;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.OnClick;


public class PartyDetailFragment extends BaseFragment {

    @Inject
    SocialRepository socialRepository;
    @Inject
    UserRepository userRepository;
    @Inject
    InventoryRepository inventoryRepository;
    @Inject
    @Named(AppModule.NAMED_USER_ID)
    String userId;

    @BindView(R.id.refreshLayout)
    SwipeRefreshLayout refreshLayout;

    @BindView(R.id.party_invitation_wrapper)
    ViewGroup partyInvitationWrapper;

    @BindView(R.id.title_view)
    TextView titleView;
    @BindView(R.id.description_view)
    TextView descriptionView;

    @BindView(R.id.new_quest_button)
    Button newQuestButton;
    @BindView(R.id.quest_detail_button)
    ViewGroup questDetailButton;
    @BindView(R.id.quest_scroll_image_view)
    SimpleDraweeView questScrollImageView;
    @BindView(R.id.quest_title_view)
    TextView questTitleView;
    @BindView(R.id.quest_participation_view)
    TextView questParticipationView;
    @BindView(R.id.quest_image_wrapper)
    ViewGroup questImageWrapper;
    @BindView(R.id.quest_image_view)
    SimpleDraweeView questImageView;
    @BindView(R.id.quest_participant_response_wrapper)
    ViewGroup questParticipantResponseWrapper;
    @BindView(R.id.quest_accept_button)
    Button questAcceptButton;
    @BindView(R.id.quest_reject_button)
    Button questRejectButton;
    @BindView(R.id.quest_progress_view)
    OldQuestProgressView questProgressView;
    @BindView(R.id.quest_participant_list)
    LinearLayout questParticipantList;


    public String partyId;
    private Group party;
    private Quest quest;
    private User user;

    @Override
    public void injectFragment(@NonNull AppComponent component) {
        component.inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);
        return inflater.inflate(R.layout.fragment_party_detail, container, false);
    }

    @Override
    public void onDestroyView() {
        socialRepository.close();
        userRepository.close();
        inventoryRepository.close();
        super.onDestroyView();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        refreshLayout.setOnRefreshListener(this::refreshParty);

        getCompositeSubscription().add(socialRepository.getGroup(partyId).subscribe(this::updateParty, RxErrorHandler.handleEmptyError()));
        getCompositeSubscription().add(userRepository.getUser(userId).subscribe(this::updateUser, RxErrorHandler.handleEmptyError()));
    }

    private void refreshParty() {
        socialRepository.retrieveGroup("party")
                .flatMap(group1 -> socialRepository.retrieveGroupMembers(group1.getId(), true))
                .subscribe(members -> {}, RxErrorHandler.handleEmptyError(), () -> {
                    if (refreshLayout != null) {
                        refreshLayout.setRefreshing(false);
                    }
                });
    }

    private void updateParty(Group party) {
        if (party == null) {
            return;
        }
        this.party = party;
        this.quest = party.getQuest();
        if (titleView == null) {
            return;
        }
        titleView.setText(party.getName());
        //TODO: FIX
        //descriptionView.setText(MarkdownParser.parseMarkdown(party.getDescription()));

        if (quest != null && !quest.getKey().isEmpty()) {
            newQuestButton.setVisibility(View.GONE);
            questDetailButton.setVisibility(View.VISIBLE);
            questImageWrapper.setVisibility(View.VISIBLE);
            Handler mainHandler = new Handler(getContext().getMainLooper());
            mainHandler.postDelayed(() -> inventoryRepository.getQuestContent(quest.getKey())
                    .firstElement()
                    .subscribe(PartyDetailFragment.this::updateQuestContent, RxErrorHandler.handleEmptyError()), 500);
        } else {
            newQuestButton.setVisibility(View.VISIBLE);
            questDetailButton.setVisibility(View.GONE);
            questImageWrapper.setVisibility(View.GONE);
            questProgressView.setVisibility(View.GONE);
        }
    }

    private void updateUser(User user) {
        if (user == null || user.getParty() == null || user.getParty().getQuest() == null) {
            return;
        }
        this.user = user;

        int invitationVisibility = View.GONE;
        if (user.getInvitations() != null && user.getInvitations().getParty() != null && user.getInvitations().getParty().getId() != null) {
            invitationVisibility = View.VISIBLE;
        }

        if (partyInvitationWrapper != null) {
            partyInvitationWrapper.setVisibility(invitationVisibility);
        }

        if (questParticipantResponseWrapper != null) {
            if (showParticipantButtons()) {
                questParticipantResponseWrapper.setVisibility(View.VISIBLE);
            } else {
                questParticipantResponseWrapper.setVisibility(View.GONE);
            }
        }

        questProgressView.configure(user);
    }

    private boolean showParticipantButtons() {
        return !(user == null || user.getParty() == null || user.getParty().getQuest() == null) && !isQuestActive() && user.getParty().getQuest().getRSVPNeeded();
    }

    private void updateQuestContent(QuestContent questContent) {
        if (questTitleView == null || !questContent.isValid()) {
            return;
        }
        questTitleView.setText(questContent.getText());
        DataBindingUtils.INSTANCE.loadImage(questScrollImageView, "inventory_quest_scroll_"+questContent.getKey());
        DataBindingUtils.INSTANCE.loadImage(questImageView, "quest_"+questContent.getKey());
        if (isQuestActive()) {
            questProgressView.setVisibility(View.VISIBLE);
            questProgressView.setData(questContent, quest.getProgress());

            questParticipationView.setText(getString(R.string.number_participants, quest.getMembers().size()));
        } else {
            questProgressView.setVisibility(View.GONE);
        }
    }

    private boolean isQuestActive() {
        return quest != null && quest.getActive();
    }

    @OnClick(R.id.new_quest_button)
    public void inviteNewQuest() {
        ItemRecyclerFragment fragment = new ItemRecyclerFragment();
        fragment.setItemType("quests");
        fragment.setItemTypeText(getString(R.string.quest));
        fragment.show(getFragmentManager(), "questDialog");
    }

    @OnClick(R.id.leave_button)
    public void leaveParty() {
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity())
                .setMessage(R.string.leave_party_confirmation)
                .setPositiveButton(R.string.yes, (dialog, which) -> socialRepository.leaveGroup(partyId)
                        .subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError())).setNegativeButton(R.string.no, (dialog, which) -> {});
        builder.show();    }

    @OnClick(R.id.quest_accept_button)
    public void onQuestAccept() {
        socialRepository.acceptQuest(user, partyId).subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());
    }


    @OnClick(R.id.quest_reject_button)
    public void onQuestReject() {
        socialRepository.rejectQuest(user, partyId).subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());
    }

    @OnClick(R.id.party_invite_accept_button)
    public void onPartyInviteAccepted() {
        if (user != null) {
            socialRepository.joinGroup(user.getInvitations().getParty().getId())
                    .subscribe(group -> {}, RxErrorHandler.handleEmptyError());
        }
    }

    @OnClick(R.id.party_invite_reject_button)
    public void onPartyInviteRejected() {
        if (user != null) {
            socialRepository.rejectGroupInvite(user.getInvitations().getParty().getId())
                    .subscribe(aVoid -> {}, RxErrorHandler.handleEmptyError());
        }
    }

    @OnClick(R.id.quest_detail_button)
    public void questDetailButtonClicked() {
        QuestDetailFragment fragment = new QuestDetailFragment();
        fragment.partyId = partyId;
        if (party != null && party.getQuest() != null) {
            fragment.questKey = party.getQuest().getKey();
        }
        if (getActivity() != null) {
            MainActivity activity = (MainActivity) getActivity();
            activity.displayFragment(fragment);
        }
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/social/party/PartyDetailFragment.kt
package com.habitrpg.android.habitica.ui.fragments.social.party

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.widget.AppCompatEditText
import androidx.core.content.ContextCompat
import androidx.core.os.bundleOf
import androidx.lifecycle.lifecycleScope
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.components.UserComponent
import com.habitrpg.android.habitica.data.ChallengeRepository
import com.habitrpg.android.habitica.data.InventoryRepository
import com.habitrpg.android.habitica.data.SocialRepository
import com.habitrpg.android.habitica.data.UserRepository
import com.habitrpg.android.habitica.databinding.FragmentPartyDetailBinding
import com.habitrpg.android.habitica.extensions.inflate
import com.habitrpg.android.habitica.helpers.HapticFeedbackManager
import com.habitrpg.android.habitica.helpers.MainNavigationController
import com.habitrpg.android.habitica.models.inventory.QuestContent
import com.habitrpg.android.habitica.models.members.Member
import com.habitrpg.android.habitica.models.social.Challenge
import com.habitrpg.android.habitica.models.social.Group
import com.habitrpg.android.habitica.models.user.User
import com.habitrpg.android.habitica.modules.AppModule
import com.habitrpg.android.habitica.ui.activities.FullProfileActivity
import com.habitrpg.android.habitica.ui.activities.MainActivity
import com.habitrpg.android.habitica.ui.fragments.BaseFragment
import com.habitrpg.android.habitica.ui.fragments.inventory.items.ItemDialogFragment
import com.habitrpg.android.habitica.ui.helpers.dismissKeyboard
import com.habitrpg.android.habitica.ui.viewHolders.GroupMemberViewHolder
import com.habitrpg.android.habitica.ui.viewmodels.PartyViewModel
import com.habitrpg.android.habitica.ui.views.HabiticaSnackbar
import com.habitrpg.android.habitica.ui.views.dialogs.HabiticaAlertDialog
import com.habitrpg.common.habitica.extensions.DataBindingUtils
import com.habitrpg.common.habitica.extensions.dpToPx
import com.habitrpg.common.habitica.extensions.loadImage
import com.habitrpg.common.habitica.helpers.ExceptionHandler
import com.habitrpg.common.habitica.helpers.launchCatching
import com.habitrpg.common.habitica.helpers.setMarkdown
import com.habitrpg.common.habitica.views.AvatarView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Named

class PartyDetailFragment : BaseFragment<FragmentPartyDetailBinding>() {

    var viewModel: PartyViewModel? = null

    override var binding: FragmentPartyDetailBinding? = null

    override fun createBinding(inflater: LayoutInflater, container: ViewGroup?): FragmentPartyDetailBinding {
        return FragmentPartyDetailBinding.inflate(inflater, container, false)
    }

    @Inject
    lateinit var challengeRepository: ChallengeRepository

    @Inject
    lateinit var socialRepository: SocialRepository

    @Inject
    lateinit var userRepository: UserRepository

    @Inject
    lateinit var inventoryRepository: InventoryRepository

    @field:[Inject Named(AppModule.NAMED_USER_ID)]
    lateinit var userId: String

    override fun injectFragment(component: UserComponent) {
        component.inject(this)
    }

    override fun onDestroyView() {
        inventoryRepository.close()
        super.onDestroyView()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding?.refreshLayout?.setOnRefreshListener { this.refreshParty() }

        binding?.questAcceptButton?.setOnClickListener { onQuestAccept() }
        binding?.questRejectButton?.setOnClickListener { onQuestReject() }
        binding?.newQuestButton?.setOnClickListener { inviteNewQuest() }
        binding?.questDetailButton?.setOnClickListener { questDetailButtonClicked() }
        binding?.leaveButton?.setOnClickListener { leaveParty() }

        binding?.invitationsView?.getLeader = null

        binding?.invitationsView?.acceptCall = {
            viewModel?.joinGroup(it) {
                lifecycleScope.launch(ExceptionHandler.coroutine()) {
                    val user = userRepository.retrieveUser(false)
                    parentFragmentManager.popBackStack()
                    MainNavigationController.navigate(
                        R.id.partyFragment,
                        bundleOf(Pair("partyID", user?.party?.id))
                    )
                }
            }
        }

        binding?.invitationsView?.rejectCall = {
            lifecycleScope.launchCatching {
                socialRepository.rejectGroupInvite(it)
                userRepository.retrieveUser(false, true)
            }
        }

        viewModel?.getGroupData()?.observe(viewLifecycleOwner) { updateParty(it) }
        viewModel?.user?.observe(viewLifecycleOwner) { updateUser(it) }
        viewModel?.getMembersData()?.observe(viewLifecycleOwner) { updateMembersList(it) }
    }

    private fun refreshParty() {
        viewModel?.retrieveGroup {
            binding?.refreshLayout?.isRefreshing = false
        }
    }

    private fun updateParty(party: Group?) {
        if (party == null) {
            return
        }
        if (binding?.titleView == null) {
            return
        }
        binding?.titleView?.text = party.name
        binding?.descriptionView?.setMarkdown(party.description)

        if (party.quest?.key?.isEmpty() == false) {
            binding?.newQuestButton?.visibility = View.GONE
            binding?.questDetailButton?.visibility = View.VISIBLE
            binding?.questImageWrapper?.visibility = View.VISIBLE
            lifecycleScope.launch(Dispatchers.Main) {
                delay(500)
                val content = inventoryRepository.getQuestContent(party.quest?.key ?: "").firstOrNull()
                if (content != null) {
                    updateQuestContent(content)
                }
            }
        } else {
            binding?.newQuestButton?.visibility = View.VISIBLE
            binding?.questDetailButton?.visibility = View.GONE
            binding?.questImageWrapper?.visibility = View.GONE
            binding?.questProgressView?.visibility = View.GONE
        }
    }

    private fun updateUser(user: User?) {
        if (user?.party?.quest == null) {
            return
        }

        var invitationVisibility = View.GONE
        if (user.invitations?.party?.id?.isNotEmpty() == true) {
            invitationVisibility = View.VISIBLE
        }

        if (binding?.partyInvitationWrapper != null) {
            binding?.partyInvitationWrapper?.visibility = invitationVisibility
        }

        if (binding?.questParticipantResponseWrapper != null) {
            if (showParticipantButtons()) {
                binding?.questParticipantResponseWrapper?.visibility = View.VISIBLE
            } else {
                binding?.questParticipantResponseWrapper?.visibility = View.GONE
            }
        }

        binding?.questProgressView?.configure(user, viewModel?.isUserOnQuest)

        if ((user.invitations?.parties?.count() ?: 0) > 0) {
            binding?.partyInvitationWrapper?.visibility = View.VISIBLE
            user.invitations?.parties?.let {
                for (invitation in it) {
                    val leaderID = invitation.inviter
                    val groupName = invitation.name

                    leaderID.let { id ->
                        lifecycleScope.launch(ExceptionHandler.coroutine()) {
                            val member = socialRepository.retrieveMember(id) ?: return@launch
                            binding?.root?.findViewById<AvatarView>(R.id.groupleader_avatar_view)
                                ?.setAvatar(member)
                            binding?.root?.findViewById<TextView>(R.id.groupleader_text_view)?.text =
                                getString(
                                    R.string.invitation_title,
                                    member.displayName,
                                    groupName
                                )
                        }
                    }

                    view?.findViewById<Button>(R.id.accept_button)?.setOnClickListener {
                        invitation.id?.let { it1 -> binding?.invitationsView?.acceptCall?.invoke(it1) }
                    }

                    view?.findViewById<Button>(R.id.reject_button)?.setOnClickListener {
                        invitation.id?.let { it1 -> binding?.invitationsView?.rejectCall?.invoke(it1) }
                    }
                }
            }
        } else {
            binding?.partyInvitationWrapper?.visibility = View.GONE
        }
    }

    private fun showParticipantButtons(): Boolean {
        return viewModel?.showParticipantButtons() ?: false
    }

    private fun updateQuestContent(questContent: QuestContent) {
        if (binding?.questTitleView == null || !questContent.isValid) {
            return
        }
        binding?.questTitleView?.text = questContent.text
        binding?.questScrollImageView?.loadImage("inventory_quest_scroll_" + questContent.key)
        if (questContent.hasGifImage()) {
            binding?.questImageView?.loadImage("quest_" + questContent.key, "gif")
        } else {
            context?.let { context ->
                DataBindingUtils.loadImage(context, "quest_" + questContent.key) {
                    if (binding?.questImageView?.drawable?.constantState != it.constantState || binding?.questImageView?.drawable == null) {
                        binding?.questImageView?.setImageDrawable(it)
                    }
                    val params = binding?.questImageView?.layoutParams ?: return@loadImage
                    params.height = it.intrinsicHeight.dpToPx(context)
                    binding?.questImageView?.layoutParams = params
                }
            }
            binding?.questImageView?.loadImage("quest_" + questContent.key)
        }
        binding?.questImageWrapper?.alpha = 1.0f
        binding?.questProgressView?.alpha = 1.0f
        context?.let { binding?.questParticipationView?.setTextColor(ContextCompat.getColor(it, R.color.text_quad)) }
        if (viewModel?.isQuestActive == true) {
            binding?.questProgressView?.visibility = View.VISIBLE
            binding?.questProgressView?.setData(questContent, viewModel?.getGroupData()?.value?.quest?.progress)

            val questParticipants = viewModel?.getGroupData()?.value?.quest?.members
            if (questParticipants?.find { it.key == userId } != null) {
                binding?.questParticipationView?.text = context?.getString(R.string.number_participants, questParticipants.size)
            } else {
                binding?.questParticipationView?.text = context?.getString(R.string.not_participating)
                context?.let { binding?.questParticipationView?.setTextColor(ContextCompat.getColor(it, R.color.red_10)) }
                binding?.questImageWrapper?.alpha = 0.5f
                binding?.questProgressView?.alpha = 0.5f
            }
        } else {
            binding?.questProgressView?.visibility = View.GONE
            val members = viewModel?.getGroupData()?.value?.quest?.members
            val responded = members?.filter { it.isParticipating != null }
            binding?.questParticipationView?.text = context?.getString(R.string.number_responded, responded?.size, members?.size)
        }
    }

    private fun updateMembersList(members: List<Member>?) {
        val leaderID = viewModel?.leaderID
        members?.forEachIndexed { index, member ->
            val memberView = (
                if ((binding?.membersWrapper?.childCount ?: 0) > index) {
                    binding?.membersWrapper?.getChildAt(index)
                } else {
                    val view = binding?.membersWrapper?.inflate(R.layout.party_member, false)
                    binding?.membersWrapper?.addView(view)
                    view
                }
                ) ?: return@forEachIndexed
            val viewHolder = GroupMemberViewHolder(memberView)
            viewHolder.bind(member, leaderID ?: "", viewModel?.user?.value?.id)
            viewHolder.onClickEvent = {
                FullProfileActivity.open(member.id ?: "")
            }
            viewHolder.sendMessageEvent = {
                member.id?.let { showSendMessageToUserDialog(it, member.displayName) }
            }
            viewHolder.transferOwnershipEvent = {
                member.id?.let { showTransferOwnerShipDialog(it, member.displayName) }
            }
            viewHolder.removeMemberEvent = {
                member.id?.let { showRemoveMemberDialog(it, member.displayName) }
            }
        }
    }

    private fun showSendMessageToUserDialog(userID: String, username: String) {
        val factory = LayoutInflater.from(context)
        val newMessageView = factory.inflate(R.layout.profile_new_message_dialog, null)

        val emojiEditText = newMessageView.findViewById<AppCompatEditText>(R.id.edit_new_message_text)

        val newMessageTitle = newMessageView.findViewById<TextView>(R.id.new_message_title)
        newMessageTitle.text = String.format(getString(R.string.profile_send_message_to), username)

        val addMessageDialog = context?.let { HabiticaAlertDialog(it) }
        addMessageDialog?.addButton(android.R.string.ok, true) { _, _ ->
            lifecycleScope.launch(ExceptionHandler.coroutine()) {
                socialRepository.postPrivateMessage(userID, emojiEditText.text.toString())
                (activity as? MainActivity)?.snackbarContainer?.let { it1 ->
                    HabiticaSnackbar.showSnackbar(
                        it1,
                        String.format(getString(R.string.profile_message_sent_to), username), HabiticaSnackbar.SnackbarDisplayType.NORMAL
                    )
                }
            }
            activity?.dismissKeyboard()
        }
        addMessageDialog?.addButton(android.R.string.cancel, false) { _, _ -> activity?.dismissKeyboard() }
        addMessageDialog?.setAdditionalContentView(newMessageView)
        addMessageDialog?.show()
    }

    private fun showTransferOwnerShipDialog(userID: String, displayName: String) {
        val dialog = context?.let { HabiticaAlertDialog(it) }
        dialog?.addButton(R.string.transfer, true) { _, _ ->
            lifecycleScope.launch(ExceptionHandler.coroutine()) {
                socialRepository.transferGroupOwnership(viewModel?.groupID ?: "", userID)
                (activity as? MainActivity)?.snackbarContainer?.let { it1 ->
                    HabiticaSnackbar.showSnackbar(
                        it1,
                        String.format(getString(R.string.transferred_ownership), displayName), HabiticaSnackbar.SnackbarDisplayType.NORMAL
                    )
                }
            }
            activity?.dismissKeyboard()
        }
        dialog?.addButton(android.R.string.cancel, false) { _, _ -> activity?.dismissKeyboard() }
        dialog?.setTitle(context?.getString(R.string.transfer_ownership_confirm))
        dialog?.setMessage(context?.getString(R.string.transfer_ownership_confirm_message, displayName))
        dialog?.show()
    }

    private fun showRemoveMemberDialog(userID: String, displayName: String) {
        val dialog = context?.let { HabiticaAlertDialog(it) }
        dialog?.addButton(R.string.remove, true) { _, _ ->
            lifecycleScope.launch(ExceptionHandler.coroutine()) {
                socialRepository.removeMemberFromGroup(viewModel?.groupID ?: "", userID)
                (activity as? MainActivity)?.snackbarContainer?.let { it1 ->
                    HabiticaSnackbar.showSnackbar(
                        it1,
                        String.format(getString(R.string.removed_member), displayName), HabiticaSnackbar.SnackbarDisplayType.NORMAL
                    )
                }
            }
            activity?.dismissKeyboard()
        }
        dialog?.addButton(android.R.string.cancel, false) { _, _ -> activity?.dismissKeyboard() }
        dialog?.setTitle(context?.getString(R.string.remove_member_confirm, displayName))
        dialog?.show()
        dialog?.show()
    }

    private fun inviteNewQuest() {
        val fragment = ItemDialogFragment()
        fragment.itemType = "quests"
        fragment.itemTypeText = getString(R.string.quest)
        fragment.isModal = true
        fragment.show(parentFragmentManager, "questDialog")
    }

    private fun getGroupChallenges(): List<Challenge> {
        val groupChallenges = mutableListOf<Challenge>()
        lifecycleScope.launchCatching {
            userRepository.getUser().collect {
                it?.challenges?.forEach { membership ->
                    val challenge = challengeRepository.getChallenge(membership.challengeID).firstOrNull()
                    if (challenge != null && challenge.groupId == viewModel?.groupID) {
                        groupChallenges.add(challenge)
                    }
                }
            }
        }
        return groupChallenges
    }

    internal fun leaveParty() {
        val context = context
        if (context != null) {
            val groupChallenges = getGroupChallenges()
            lifecycleScope.launch(Dispatchers.Main) {
                delay(500)
                if (groupChallenges.isNotEmpty()) {
                    val alert = HabiticaAlertDialog(context)
                    alert.setTitle(R.string.party_challenges)
                    alert.setMessage(R.string.leave_party_challenges_confirmation)
                    alert.addButton(R.string.keep_challenges, true) { _, _ ->
                        viewModel?.leaveGroup(groupChallenges, true) {
                            parentFragmentManager.popBackStack()
                            MainNavigationController.navigate(R.id.noPartyFragment)
                        }
                    }
                    alert.addButton(R.string.leave_challenges_delete_tasks, false, isDestructive = true) { _, _ ->
                        viewModel?.leaveGroup(groupChallenges, false) {
                            parentFragmentManager.popBackStack()
                            MainNavigationController.navigate(R.id.noPartyFragment)
                        }
                    }
                    alert.setExtraCloseButtonVisibility(View.VISIBLE)
                    alert.show()
                } else {
                    val alert = HabiticaAlertDialog(context)
                    alert.setTitle(R.string.leave_party_confirmation)
                    alert.setMessage(R.string.rejoin_party)
                    alert.addButton(R.string.leave, isPrimary = true, isDestructive = true) { _, _ ->
                        viewModel?.leaveGroup(groupChallenges, false) {
                            parentFragmentManager.popBackStack()
                            MainNavigationController.navigate(R.id.noPartyFragment)
                        }
                    }
                    alert.setExtraCloseButtonVisibility(View.VISIBLE)
                    alert.show()
                }
            }
        }
    }

    private fun onQuestAccept() {
        HapticFeedbackManager.tap(requireView())
        viewModel?.acceptQuest()
    }

    private fun onQuestReject() {
        HapticFeedbackManager.tap(requireView())
        viewModel?.rejectQuest()
    }

    private fun questDetailButtonClicked() {
        MainNavigationController.navigate(PartyFragmentDirections.openQuestDetail())
    }
}
